services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/ApplicationData.kt

```diff
@@ -36,6 +36,20 @@ sealed interface ApplicationData {
     fun copy(pendingOnClient: Boolean): ApplicationData
     fun toAlluData(hankeTunnus: String): AlluApplicationData
     fun customersWithContacts(): List<CustomerWithContacts>
+
+    /**
+     * Returns a set of email addresses from customer contact persons that:
+     * - are not null, empty or blank.
+     * - do not match the optional [omit] argument.
+     */
+    fun contactPersonEmails(omit: String? = null): Set<String> =
+        customersWithContacts()
+            .flatMap { customer ->
+                customer.contacts.mapNotNull {
+                    if (it.email.isNullOrBlank() || it.email == omit) null else it.email
+                }
+            }
+            .toSet()
```

Comment:
I guess it does look a bit messy, but I wanted to avoid multiple iterations over the (basically) same collection here. I'll see what I can do.

Reply:
You could use a sequence to avoid the multiple iterations: https://kotlinlang.org/docs/sequences.html

But I don't think performance is an issue here with any of these implementations.

Reply:
How about now?

I thought of sequences here but I imagine the collections are too small to gain anything from it.

Reply:
I think there's still some unnecessary nesting. The function shouldn't be any more complicated than it needs to be.

Reply:
I have to say I disagree. I don't see the issue in why 

`.mapNotNull { contact -> contact.email.takeUnless { it.isNullOrBlank() } }`

is more complicated than 
```
            .mapNotNull { it.email }
            .filter { it.isNotBlank() }
```

Are we perhaps debating a matter of personal preference?

Reply:
My reasoning for this solution:
1. No reason to use a sequence.
2. I don't want extra iterations over the collection.

As I don't think this line is complicating things (readability etc.) it is the best solution I've got.
`.mapNotNull { contact -> contact.email.takeUnless { it.isNullOrBlank() } }`

Reply:
>My reasoning for this solution:
>1. No reason to use a sequence.
>2. I don't want extra iterations over the collection.

These, however, are personal preferences unless there are performance concerns to justify them.

If the collection is small enough that it's not worth the trouble to use a sequence (adding one `.asSequence()` line), then they are certainly small enough that one more iteration over them won't make any measurable difference. And it's not certain that the extra iteration would be more expensive than the added function calls in the inner function of the nested solution.

Reply:
I'm not saying that the current implementation is especially hard to read, or hard at all. But the small complication it has is just very unnecessary.

Reply:
> won't make any measurable difference. 

This pretty much sums up our discussion. Complexity wise, there is no meaningful / significant difference either. I do not wan't to get stuck here, so I changed the implementation as suggested.